crates/ty_python_semantic/resources/mdtest/type_compendium/tuple.md

```diff
@@ -0,0 +1,122 @@
+# Tuples
+
+## Tuples as product types
+
+Tuples can be used to construct product types. Inhabitants of the type `tuple[P, Q]` are ordered
+pairs `(p, q)` where `p` is an inhabitant of `P` and `q` is an inhabitant of `Q`, analogous to the
+Cartesian product of sets.
+
+```py
+from typing_extensions import assert_type
+
+class P: ...
+class Q: ...
+
+def _(p: P, q: Q):
+    assert_type((p, q), tuple[P, Q])
+```
+
+## Subtyping relationships
+
+The type `tuple[S1, S2]` is a subtype of `tuple[T1, T2]` if and only if `S1` is a subtype of `T1`
+and `S2` is a subtype of `T2`, and similar for other lengths of tuples:
+
+```py
+from ty_extensions import static_assert, is_subtype_of
+
+class T1: ...
+class S1(T1): ...
+class T2: ...
+class S2(T2): ...
+
+static_assert(is_subtype_of(tuple[S1], tuple[T1]))
+static_assert(not is_subtype_of(tuple[T1], tuple[S1]))
+
+static_assert(is_subtype_of(tuple[S1, S2], tuple[T1, T2]))
+static_assert(not is_subtype_of(tuple[T1, S2], tuple[S1, T2]))
+static_assert(not is_subtype_of(tuple[S1, T2], tuple[T1, S2]))
+```
+
+Different-length tuples are not related via subtyping:
+
+```py
+static_assert(not is_subtype_of(tuple[S1], tuple[T1, T2]))
+```
+
+## The empty tuple
+
+The type of the empty tuple `()` is spelled `tuple[()]`. It is *not* a singleton type, as it can be
+subclassed:
```

Comment:
I think the empty tuple would not be a singleton even if we didn't model the possibility of tuple subclasses. There was strong consensus in https://discuss.python.org/t/should-we-specify-in-the-language-reference-that-the-empty-tuple-is-a-singleton/67957 that empty tuple being a singleton is an implementation detail of CPython, not part of the language.

I think disjointness of heterogenous tuple types with other instance types might be an area where we do take into account the possibility of subclasses?

(And it's worth noting that our current treatment of the empty tuple as `AlwaysFalsy` -- mentioned in that doc -- is inconsistent with modeling the possibility of tuple subclasses.)

Reply:
> I think the empty tuple would not be a singleton even if we didn't model the possibility of tuple subclasses. There was strong consensus in https://discuss.python.org/t/should-we-specify-in-the-language-reference-that-the-empty-tuple-is-a-singleton/67957 that empty tuple being a singleton is an implementation detail of CPython, not part of the language.

Thanks, updated and added that reference.



> I think disjointness of heterogenous tuple types with other instance types might be an area where we do take into account the possibility of subclasses?

Correct. I added a test for this.

> (And it's worth noting that our current treatment of the empty tuple as `AlwaysFalsy` -- mentioned in that doc -- is inconsistent with modeling the possibility of tuple subclasses.)

:+1: